articles/contributing/api-design-principles.md

```diff
@@ -0,0 +1,426 @@
+
+    - In cases where an operation type with three qubit array
+        inputs is commonly taken as an input or returned as an
+        output, providing a UDT such as
+        `newtype TimeDependentBlockEncoding = ((Qubit[], Qubit[], Qubit[]) => Unit is Adj + Ctl)`
+        can help provide a useful shorthand.
+
+    - **DO** introduce new user-defined types to indicate that a given
+        base type should only be used in a very particular sense. For
+        example:
+
+      - An operation that should be interpreted specifically as an
+          operation that encodes classical data into a quantum
+          register may be appropriate to label with a user-defined
+          type `newtype InputEncoder = (Apply : (Qubit[] => Unit))`.
+
+    - **DO** introduce new user-defined types with named items that
+        allow for future extensibility (e.g.: a results structure that
+        may contain additional named items in the future). For example:
+
+      - When an operation `TrainModel` exposes a large number of
+          configuration options, exposing these options as a new
+          `TrainingOptions` UDT and providing a new function
+          `DefaultTrainingOptions : Unit -> TrainingOptions` allows
+          users to override specific named items in TrainingOptions
+          UDT values while still allowing library developers to add
+          new UDT items as appropriate.
+
+    - **DO** declare named items for new user-defined types in
+        preference to requiring users to know the correct tuple
+        deconstruction. For example:
+
+      - When representing a complex number in its polar
+          decomposition, prefer
+          `newtype ComplexPolar = (Magnitude: Double, Argument: Double)` to
+          `newtype ComplexPolar = (Double, Double)`.
+
+- **DON'T** use user-defined types in ways that increase cognitive
+    load or that require the user to learn additional concepts and
+    nomenclature.
+
+  - **DON'T** introduce user-defined types that require the user to
+      make frequent use of the unwrap operator (`!`), or that commonly
+      require multiple levels of unwrapping. Possible mitigation strategies
+      include:
+
+    - When exposing a user-defined type with a single item,
+        consider defining a name for that item. For instance,
+        consider `newtype Encoder = (Apply : (Qubit[] => Unit is Adj + Ctl))`
+        in preference to `newtype Encoder = (Qubit[] => Unit is Adj + Ctl)`.
+
+    - Ensuring that other functions and operations can accept
+        "wrapped" UDT instances directly.
+
+  - **DON'T** introduce new user-defined types that duplicate
+      built-in types without providing additional expressiveness. For
+      example:
+
+    - A UDT `newtype QubitRegister = Qubit[]` provides no
+        additional expressiveness over `Qubit[]`, and is thus harder
+        to use with no discernable benefit.
+
+    - A UDT `newtype LittleEndian = Qubit[]` documents how the
```

Comment:
How does this compare to using types to restrict use of variables? E.g., see Python typing's newtype docs https://docs.python.org/3/library/typing.html#newtype

Reply:
That's a great question, thanks for raising it! Whereas Python's `NewType` function creates a subtype of a given type, whereas the Q# `newtype` keyword creates a new type that wraps a given type. Thus, while a Python value of type `NewType('UserId', int)` can be used anywhere a value of type `int` can be used, with the declaration `newtype UserId = Int`, a Q# value of type `UserId` cannot be used in place of an `Int` without using the unwrap operator (`!`). From that perspective, the API design principles around UDTs look slightly different than around Python `NewType`s.

@KittyYeungQ, would it perhaps make sense to add a comparison to Python's `NewType` function somewhere in the docs?

Reply:
Thanks for the explanation. But I have to ask, is there a reason for a newtype to not be a subclass for what type it aliases to? If you have to unwrap a type every time to do basic operations, wouldn't that hamper code reusability?
On the other hand, I can see why, if you were to act on e.g. a Complex type (https://docs.microsoft.com/en-us/quantum/language/type-model?view=qsharp-preview#user-defined-types) with a map, and it would apply to the real and imaginary part as if it was a tuple, that would be weird.

Reply:
@bettinaheim could answer better than I could, but the rough version was that we actually tried making UDTs subtypes of their base types (that was the case until Q# 0.3, if I recall), and that actually wound up pretty confusing the way that UDTs are used in practice. Especially with singleton–tuple equivalence, which has turned out to be a very useful feature across the language, it wasn't always clear when to think of a value as being of the new subtype and when to think of it as being implicitly downcast to its base type.

The way we're currently using singleton UDTs like `newtype LittleEndian = Qubit[]` in the libraries, they mark that UDT is opaque and is intended for use with operations that are specifically know how to handle the interpretation of that base type. For example, SWAPing the order of qubits in a `LittleEndian` violates the invariant that it should be interpreted as a little-endian integer, and so we don't want to accidentally pass a LittleEndian to something like SwapReverseRegister unless we explicitly mean to do that.

Reply:
Certainly, in the API, there will be some assumption on the endianness of the input? Now, this does not have to be enforced through a type or is there some kind of Register type that has a defined ordering? How will you otherwise enforce the input to some quantum algorithm adheres to the assumptions?

Reply:
Not sure if I understand the questions in the last reply properly. To follow up on that udts not being subtypes of their underlying type, in addition to what @cgranade already said: If they were considered subtypes then they would cast automatically if needed, which would encourage relying on their shape (meaning relying on the exact items they contain). Having them being their own type encourages instead relying on constructs like complex::Re to access items, which makes it easier to add additional items in the future without breaking code. Of course, all the constructs that use the unwrap operator ! will break when additional items are added to a type, and in that sense should be avoided if possible. Constructs using ! rely heavily on the exact version of that type, and should ideally only be used where the type is declared to define functions that handle certain functionalities for that type.